ttnn/cpp/ttnn/operations/uniform/device/kernels/writer_uniform.cpp

```diff
@@ -0,0 +1,59 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "common/constants.hpp"
+#include "dataflow_api.h"
+#include "debug/dprint.h"
+
+using namespace tt;
+
+void kernel_main() {
+    constexpr uint32_t intermed_cb_id = get_compile_time_arg_val(0);
+    constexpr uint32_t dst_cb_id = get_compile_time_arg_val(1);
+    constexpr bool output_is_dram = get_compile_time_arg_val(2) == 1;
+
+    uint32_t dst_addr = get_arg_val<uint32_t>(0);
+    uint32_t start_id = get_arg_val<uint32_t>(1);
+    uint32_t num_tiles = get_arg_val<uint32_t>(2);
+    uint32_t end_id = start_id + num_tiles;
+
+    const InterleavedAddrGenFast<output_is_dram> output_addrg = {
+        .bank_base_address = dst_addr, .page_size = get_tile_size(dst_cb_id), .data_format = get_dataformat(dst_cb_id)};
+
+    cb_reserve_back(dst_cb_id, 1);
+    uint32_t dst_cb_write_ptr = get_write_ptr(dst_cb_id);
+
+    for (uint32_t i = start_id; i < end_id; ++i) {
+        cb_wait_front(intermed_cb_id, 1);
+
+        uint32_t intermed_cb_read_ptr = get_read_ptr(intermed_cb_id);
+        auto intermed_cb_addr = reinterpret_cast<float *>(intermed_cb_read_ptr);
+
+#ifdef OUTPUT_DTYPE_FLOAT32
+        noc_async_write_tile(i, output_addrg, intermed_cb_read_ptr);
+        noc_async_write_barrier();
+        cb_pop_front(intermed_cb_id, 1);
+#endif
+
+#ifdef OUTPUT_DTYPE_BFLOAT16
+        auto dst_cb_addr = reinterpret_cast<uint8_t *>(dst_cb_write_ptr);
```

Comment:
Is this step not possible on the SFPU I mean conversion from fp32 to bfp16  using SFPSTORE instruction mod 2 ?

Reply:
I can't as I have to use vFloat after TTI_SFPSTORE (due to the issue described [in this cmt](https://github.com/tenstorrent/tt-metal/pull/13735#discussion_r1810535669)).

Reply:
I tried mod 2, the output tile look likes this: 512 numbers in range (0, 1), 512 zeros. Do you have any idea why?

<img width="1836" alt="image" src="https://github.com/user-attachments/assets/3193ca58-dfa4-416e-9113-dab0d43459e7">

Also, I found no ckernel example use mode 2. Found no ckernel file handle 2 dtypes, neither.

Reply:
Do you change the type of your circular buffer depending on whether you are writing fp32 of bfp16 ?

Reply:
Yes, I do. Please check this [draft commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/71a1528d602ce45c5872ffcb67f7c8ba8400ae4c).

Reply:
Did it work in your draft commit ?

Reply:
No, it didn't. Half of the output are zero as the image above.

Reply:
If you use fp32 and use the other TTI_SFPSTORE  with mod 0 or 3 do you get the full tile ?

Reply:
Here is what I've tried
- mod = 2: get half bfloat16 tile.
- mod = 4: get full float32 tile.
- mod = 0: get full float32, bfloat16 tile. However, there is a precision error with bfloat16 which may cause the generated number to be equal to `to` param.

Reply:
Okay for bfloat16 can you use mod 0 but add another instruction to zero out the last 16 bits and see if you see the error. I will look a bit more into it to see how we can properly use mod=2.

Reply:
If you are not out of registers you can 
TT_SFPLOADI(p_sfpu::LREG1, 8, 0XFFFF);
and then do the and  with the register containing the high 16 bits.

Reply:
or     
    TT_SFPLOADI(p_sfpu::LREG1, 10, 0);
    TT_SFPLOADI(p_sfpu::LREG1, 8, 0XFFFF);

Reply:
I tried zero-out last 16 bits, the precision err remains, still. 

Another issue with mod 0 or 2, with bfloat16 is that there are more chances random number are duplicate (which idk why). In an input tensor of [1024, 1024] shape, there are only 268 distinct elements whereas with the initial approach of casting to bfloat16 in writer kernel we could reach more than 30000 distinct elements.

If there is no update, I think I'll stick with the initial approach of casting to bfloat16 in the writer kernel.

Reply:
Ok that sounds good.

Reply:
The SFPU rounds to the nearest even, so some rounding issue is expected. But I was hoping zeroing out the last 16 bits would help.

Reply:
But I meant zeroing out the last 16 bits of the random number between 0-1 though. Not the final result.

Reply:
More duplicate random numbers, I think makes sense, as we are taking only a small fraction of the random bit pattern , so there are more duplicate cases, but when you use the full range in fp32 that is 23 bits of randomness, after multiplication and addition you have more uniqueness. So I guess your approach would fare well in terms of randomness.

Reply:
I've rollbacked the code to the initial approach. Please take another look. Thank you.

Reply:
Thanks a lot @BuiChiTrung  I will take a look , but could you tell me how I can reproduce the error you mentioned when you used the SFPSTORE for fp16 ? We might want to find out a workaround and such a test might help us not ignore the error. Thanks.

Reply:
You can checkout this [commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/e2d9767ad79c03818b21dcbc48fbb6db934783e8), turn on the DPRINT, run the unit-test to see a tile of bfp16 is printed with 512 zeros.

Reply:
I meant the error that it reaches the parameter ```to```,  is it triggered by the existing test cases you added ?

Reply:
Oh, for that bug. You can change SFPSTORE instr_mod to 0, use the same test but with bigger shape [100, 100].

So can I merge the code using the current approach? If so, please approve this PR.

Reply:
@BuiChiTrung when you said when you used SFPSTORE mod 0 and bfp16, you reach the TO parameter. Did you reach it when the range was all negative ? like [-200, -100) or did you reach it with a positive TO. If you reached it with a positive TO then we need to investigate it since it is not supposed to happen.

Reply:
And did you create the bug ? Could you share the issue number ?

Reply:
TO is reached when it's positive. I haven't tried with negative range.

Reply:
FYI, the sfpu API create random number in range `[from, from + scale]`. To make sure that generated number < `to` (from +scale), in the uniform program factory, I minus to with an epsilon: `-1e6`. 
```
const float eps = 1e-6;
        union {
            float f;
            uint32_t u;
        } f2u_from, f2u_to;
        f2u_from.f = operation_attributes.from;
        f2u_to.f = operation_attributes.to - eps;  // -eps make sure that generated number is < operation_attributes.to
```

But with mod 0, `to` is reached.